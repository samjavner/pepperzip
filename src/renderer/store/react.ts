import * as React from "react";
import { Commands, Dispatch, Store, Update, UseStoreOptions } from "./";

export interface Msg<M extends object, U extends Update<M>> {
    type: keyof U;
    payload?: any;
}

export function useLocalStore(log: (msg: any) => void = () => ({})) {
    return function actual<
        M extends object,
        U extends Update<M>,
        C extends Commands = {}
    >({
        init,
        update,
        createCommands,
        memo = [],
    }: UseStoreOptions<M, U, C>): Store<M, U, C> {
        const [model, reducerDispatch] = React.useReducer(
            (state: M, action: Msg<M, U>) =>
                update[action.type](state, action.payload),
            init
        );

        const vm = React.useMemo(
            () => {
                const dispatch: Dispatch<U> = {} as any;
                Object.keys(update).forEach(type => {
                    (dispatch as any)[type] = (payload: any) => {
                        const msg =
                            update[type].length === 1
                                ? { type }
                                : { type, payload };
                        log(msg);
                        reducerDispatch(msg);
                    };
                });
                const commands: C = createCommands
                    ? createCommands(model, dispatch)
                    : (({} as any) as C);
                return {
                    ...model,
                    ...dispatch,
                    ...commands,
                };
            },
            [model, update, ...memo]
        );

        return vm;
    };
}
