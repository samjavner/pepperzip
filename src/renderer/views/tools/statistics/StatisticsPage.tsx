import * as React from "react";
import { connect } from "react-redux";
import { GlobalState } from "../../../core/model";

export const StatisticsPage: React.SFC = props => <div>Statistics</div>;

const mapStateToProps = (state: GlobalState) => ({});

export default connect(mapStateToProps)(StatisticsPage);
