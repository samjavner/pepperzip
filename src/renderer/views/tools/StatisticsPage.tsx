import * as React from "react";
import { Statistics } from "../../core/tools/statistics";

export const StatisticsPage: React.FunctionComponent<
    Statistics
> = statistics => (
    <div>
        <button className="button" onClick={statistics.roll}>
            {statistics.dieFace}
        </button>
    </div>
);

export default StatisticsPage;
