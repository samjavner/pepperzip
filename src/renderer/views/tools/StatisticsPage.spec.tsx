import * as React from "react";
import * as renderer from "react-test-renderer";
import { StatisticsPage } from "./StatisticsPage";

describe("StatisticsPage", () => {
    it("renders a button with the current dieFace", () => {
        const component = renderer.create(
            <StatisticsPage
                dieFace={5}
                newFace={() => ({})}
                roll={() => Promise.resolve()}
            />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
