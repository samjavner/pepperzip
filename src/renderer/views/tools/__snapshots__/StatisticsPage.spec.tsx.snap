// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticsPage renders a button with the current dieFace 1`] = `
<div>
  <button
    className="button"
    onClick={[Function]}
  >
    5
  </button>
</div>
`;
